kernel/src/capability_ptr.rs

```diff
@@ -0,0 +1,135 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Google LLC 2024.
+
+//! Defines the CapabilityPtr type
+
+use core::fmt::{Formatter, LowerHex, UpperHex};
+use core::ops::AddAssign;
+
+/// A pointer to userspace memory.
+///
+/// A [CapabilityPtr] points to memory a userspace processes may be
+/// permitted to read, write, or execute. It is sized exacly to a
+/// register that can pass values between userspace and kernel and at
+/// least the size of a word ([usize]) [^note1]. Operations on the
+/// pointer may affect permissions, e.g. offsetting the pointer beyond
+/// the bounds of the memory object invalidates it. Like a `*const
+/// ()`, a [CapabilityPtr] may also "hide" information by storing a
+/// word of data with no memory access permissions.
+///
+/// [CapabilityPtr] should be used to store or pass between kernel and
+/// userspace a value that may represent a valid userspace reference.
+///
+/// [^note1]: Depending on the architecutre, this the size of a
+/// [CapabilityPtr] may be a word size or larger, e.g., if registers
+/// can store metadata such as access permissions.
```

Comment:
I think we should mention CHERI somewhere in this, because I think `CapabilityPtr` will be hard to understand for someone who is isn't very familiar with CHERI. `[^note1]` seems like a good place to mention this, as it's already giving an example of a place where `CapabilityPtr` is useful.

Reply:
Will add comment when CHERI lands, that this always wraps a CHERI capability on that platform